### PR TITLE
Loosen dependency on jemalloc

### LINKED ISSUE #1442: RocksDB shouldn't determine at build time whether to use jemalloc / tcmalloc
RocksDB (via `build_tools/build_detect_platform`) determines at build time whether to use jemalloc, tcmalloc, or none. This is wrong for two reasons:
1. The choice of memory allocator should be left to the main program, not a library, so RocksDB shouldn't add `-ljemalloc` (or `-ltcmalloc`) to its own link line. Until recently, Linux would propagate a shared library's DT_NEEDED entries to the parent (see `--add-needed` or `--copy-dt-needed-entries` in `man ld`), so RocksDB's actions would force the main program to use jemalloc / tcmalloc. If we want to use `jemalloc` / `tcmalloc` in RocksDB **executables** (such as `ldb` and `db_bench`), then `-ljemalloc` / `-ltcmalloc` should be added to those executables' link lines.
2. If RocksDB determines that `jemalloc` is available, it assumes it will only be linked into binaries that use it as its memory allocator, and thus it calls `jemalloc`-specific functions, see the call to `malloc_stats_print` in https://github.com/facebook/rocksdb/blob/c90c48d3c89141bc8839a0fa69a6b5b2d4f03d08/db/db_impl.cc#L599. This leads to a link error if you try to link RocksDB into an executable that doesn't use `jemalloc`. The proper thing to do is to declare `malloc_stats_print` as weak, and check for its existence at runtime, but even that's technically not enough -- it should check that `jemalloc` is actually in use. I solved this in `folly` in a horrible way that I'm not at all proud of: https://github.com/facebook/folly/blob/1a48bcd9c77353535820239dff3e7f5f7c1e93bb/folly/Malloc.h#L143

### PR DESCRIPTION
**Pull request by tudor**
Fixes #1442 and #1717.

No longer add -ljemalloc to library link lines.

No longer call malloc_print_stats() unless we're actually using jemalloc (use weak linking).

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
@tudor We are getting the following error in our contbuild. Do you mind fixing it? Thanks. https://gist.github.com/yiwu-arbug/aca8bc19c11316d73990d70076a168a4

**Comment by tudor** [comment]
@yiwu-arbug By "our" contbuild, I'm assuming you mean Facebook's? I'm no longer a Facebook employee. Can you check whether the contbuild sets the JEMALLOC_INCLUDE environment variable explicitly? (I removed that from build_detect_platform, I can add it back if that's the reason.)

**Comment by tudor** [comment]
Ah, you do mean Facebook's contbuild, and it does set it.

`build_tools/fbcode_config.sh:JEMALLOC_INCLUDE=" -I $JEMALLOC_BASE/include/"`

**Comment by tudor** [comment]
@yiwu-arbug Try now?

**Comment by yiwu-arbug** [comment]
@tudor Thanks. Let's wait for the build.

File: Makefile [hunk 1]
```diff
@@ -197,12 +197,18 @@ endif
 
 ifndef DISABLE_JEMALLOC
 	ifdef JEMALLOC
-		PLATFORM_CXXFLAGS += "-DROCKSDB_JEMALLOC"
-		PLATFORM_CCFLAGS +=  "-DROCKSDB_JEMALLOC"
+		PLATFORM_CXXFLAGS += -DROCKSDB_JEMALLOC
+		PLATFORM_CCFLAGS +=  -DROCKSDB_JEMALLOC
 	endif
-	EXEC_LDFLAGS := $(JEMALLOC_LIB) $(EXEC_LDFLAGS)
+	# JEMALLOC_INCLUDE and JEMALLOC_LIB are forced in the Facebook
```
Comment (IslamAbdelRahman) [inline Makefile]: This comment is confusing me, This Makefile is not Facebook specific.

Comment (yiwu-arbug) [inline Makefile]: Hi @tudor. Do you mind updating the comment? Thanks!

**Pull request closed without merging**